### PR TITLE
Splitattrs equalise

### PR DESCRIPTION
**Pull request by pp-mo**
Addresses #4988 
---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by pp-mo** [comment]
https://github.com/SciTools/iris/pull/5586/commits/3cb375e7da4a6c993ca696a1f89ae1cd6e49d183 at this point, demonstrated expected-failures
(see [test log](https://github.com/SciTools/iris/actions/runs/6882661416/job/18721610409?pr=5586) )

**Comment by pp-mo** [comment]
@trexfeathers are you interested in this ?

We still have the little thing to fix merge errors in #5585, 
and the merge-back #5152,
but https://github.com/orgs/SciTools/projects/5 is nearly nearly there !

File: lib/iris/util.py [hunk 1]
```diff
@@ -2059,31 +2059,53 @@ def equalise_attributes(cubes):
...
+    from iris.cube import CubeAttrsDict
+
+    cube_attrs = [cube.attributes for cube in cubes]
     removed = []
+
+    # *IF* they are cube attributes, then convert to "paired-keys" form.
+    is_split_dicts = any(
+        isinstance(attrs, CubeAttrsDict) for attrs in cube_attrs
+    )
+    if is_split_dicts:
```
Comment (trexfeathers) [inline lib/iris/util.py:2075]: This is already handled by your addition to `_convert_splitattrs_to_pairedkeys_dict()`:

https://github.com/SciTools/iris/blob/35ec095a6293394c599c71a8238db5f10e2f9e13/lib/iris/common/_split_attribute_dicts.py#L39-L41

So can we get away without any checks at either end?

File: lib/iris/util.py [hunk 2]
```diff
@@ -2059,31 +2059,53 @@ def equalise_attributes(cubes):
...
-                and np.all(cube.attributes[key] == cubes[0].attributes[key])
-            )
+            if (key in cube_keys and np.all(attrs[key] == cube_attrs[0][key]))
         ]
     keys_to_remove.difference_update(common_keys)
 
-    # Remove all the other attributes.
+    # rework the keys list if we were handling split-attributes.
+    if is_split_dicts:
+        # Collect all the ('local'/'global, name) keys + retain just the names.
```
Comment (trexfeathers) [inline lib/iris/util.py:2097]:
```suggestion
        # Collect all the ('local'/'global', name) keys + retain just the names.
```

File: lib/iris/util.py [hunk 3]
```diff
@@ -2059,31 +2059,53 @@ def equalise_attributes(cubes):
...
+    # deferred imports to avoid circularity problem
+    from iris.common._split_attribute_dicts import (
+        _convert_splitattrs_to_pairedkeys_dict,
+    )
+    from iris.cube import CubeAttrsDict
+
+    cube_attrs = [cube.attributes for cube in cubes]
     removed = []
+
+    # *IF* they are cube attributes, then convert to "paired-keys" form.
```
Comment (trexfeathers) [inline lib/iris/util.py:2071]: Can this comment be enhanced to explain why the rest of the code 'magically works' once the keys are paired keys?

File: lib/iris/tests/unit/util/test_equalise_attributes.py [hunk 4]
```diff
@@ -153,5 +158,111 @@ def test_complex_somecommon(self):
...
 
+class TestSplitattributes:
+    """
+    Extra testing for cases where attributes differ specifically by type
+
+    That is, where there is a new possibility of 'mismatch' due to the newer "typing"
+    of attributes as global or local.
+
+    Specifically, it is now possible that although
+    "cube1.attributes.keys == cube2.attributes.keys()",
```
Comment (trexfeathers) [inline lib/iris/tests/unit/util/test_equalise_attributes.py:169]:
```suggestion
    "cube1.attributes.keys() == cube2.attributes.keys()",
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]

File: lib/iris/util.py (hunk 1 above)
Comment (pp-mo) [reply lib/iris/util.py:2075]: The original point here was the the conversion to/from paired-keys type is only needed for split-type attributes
(which technically we should _always_ have, if it is only used on cubes, but I still think it's safer to not insist on that)

On reflection it should still _work_, if less efficient, to do this with regular attributes dictionaries too -- i.e. always + regardless of type.
I will try that, as it clearly would make the code simpler.

Comment (pp-mo) [reply lib/iris/util.py]: Yup, it works !
https://github.com/SciTools/iris/pull/5586/commits/c3e2175b7ea7735ed8ab2b49d926d70922a4afde

File: lib/iris/util.py (hunk 3 above)
Comment (pp-mo) [reply lib/iris/util.py]: I've tried now https://github.com/SciTools/iris/pull/5586/commits/c3e2175b7ea7735ed8ab2b49d926d70922a4afde
Please see what you think !

**Comment by pp-mo** [comment]
Thanks @trexfeathers I think the latest addresses your points so far.
Please take a look + see what you think now.

**Review by trexfeathers — APPROVED** [review]

**Pull request merged**